### PR TITLE
[MLAS AArch64] SQNBitGemm CompInt8 kernel

### PR DESCRIPTION
**Pull request by edgchen1**
# Description

Implement ARM NEON SQNBitGemm kernel that first block quantizes A to int8 and then does int8 multiplication.

## Benchmarks
Note: `ComputeType:1` is original float multiplication implementation, `ComputeType:4` is new int8 multiplication implementation.

Benchmark runs from 2337375956.

name|median_real|median_cpu
-|-|-
SQNBITGEMM<4>/BlkLen:16/M:1/N:4096/K:11008/Threads:8/Symmetric:1/ComputeType:1/real_time|1700949 ns|1700130 ns
SQNBITGEMM<4>/BlkLen:16/M:1/N:4096/K:11008/Threads:8/Symmetric:1/ComputeType:4/real_time|1154813 ns|1150777 ns
SQNBITGEMM<4>/BlkLen:256/M:1/N:4096/K:11008/Threads:8/Symmetric:1/ComputeType:1/real_time|1530306 ns|1535207 ns
SQNBITGEMM<4>/BlkLen:256/M:1/N:4096/K:11008/Threads:8/Symmetric:1/ComputeType:4/real_time|706529 ns|697831 ns
SQNBITGEMM<4>/BlkLen:16/M:1/N:11008/K:4096/Threads:8/Symmetric:1/ComputeType:1/real_time|1747413 ns|1744727 ns
SQNBITGEMM<4>/BlkLen:16/M:1/N:11008/K:4096/Threads:8/Symmetric:1/ComputeType:4/real_time|1141534 ns|1143927 ns
SQNBITGEMM<4>/BlkLen:256/M:1/N:11008/K:4096/Threads:8/Symmetric:1/ComputeType:1/real_time|1581004 ns|1562500 ns
SQNBITGEMM<4>/BlkLen:256/M:1/N:11008/K:4096/Threads:8/Symmetric:1/ComputeType:4/real_time|741127 ns|735486 ns
SQNBITGEMM<4>/BlkLen:16/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:1/real_time|618847 ns|617318 ns
SQNBITGEMM<4>/BlkLen:16/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:4/real_time|277728 ns|276930 ns
SQNBITGEMM<4>/BlkLen:256/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:1/real_time|570383 ns|558036 ns
SQNBITGEMM<4>/BlkLen:256/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:4/real_time|204531 ns|200321 ns

# Motivation and Context

ARM NEON SQNBitGemm kernel performance.

### CODE REVIEW

File: onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp [hunk 1]
```diff
@@ -448,42 +414,439 @@ MlasQNBitBlkDequantBForSgemmNeon(
...
+            UnrolledLoop<NCols>([&](size_t i) {
+                const int8x16_t zp_v = vdupq_n_s8(b_zp[i]);
+                bv[i] = vsubq_s8(bv[i], zp_v);
+            });
+
+            // compute quantized dot product
+            int32x4_t dot[NCols];
+            UnrolledLoop<NCols>([&](size_t i) {
+                const int32x4_t zero_v = vdupq_n_s32(0);
+                dot[i] = vdotq_s32(zero_v, av, bv[i]);
```
Comment (edgchen1) [inline onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp:671]: need to check if ARM NEON dot instructions are available

File: cmake/onnxruntime_mlas.cmake [hunk 2]
```diff
@@ -348,6 +356,8 @@ else()
           ${MLAS_SRC_DIR}/qgemm_kernel_sdot.cpp
           ${MLAS_SRC_DIR}/sqnbitgemm_kernel_neon.cpp
         )
+        set_source_files_properties(${MLAS_SRC_DIR}/sqnbitgemm_kernel_neon.cpp
+                                    PROPERTIES COMPILE_FLAGS " -march=armv8.2-a+dotprod")
```
Comment (yufenglee) [inline cmake/onnxruntime_mlas.cmake:360]: we still need to support the fp32+int4 on lower arch. Will this break the case?

Comment (yufenglee) [reply cmake/onnxruntime_mlas.cmake:360]: I see you have guard to only run int4 for armv8.2 and above. It is fine to have this assumption.


        // MlasSQNBitGemmDispatchNeon has a dependency on dot product instructions
        this->SQNBitGemmDispatch = &MlasSQNBitGemmDispatchNeon;

File: onnxruntime/test/mlas/unittest/test_sqnbitgemm.cpp [hunk 3]
```diff
@@ -95,10 +194,11 @@ class MlasSQNBitGemmTest : public MlasTestBase {
 
  public:
   void Test(size_t M, size_t N, size_t K,
+            MLAS_SQNBIT_GEMM_COMPUTE_TYPE ComputeType,
             bool WithBias, bool Symmetric, bool WithThreadpool) {
     MLAS_THREADPOOL* Threadpool = WithThreadpool ? GetMlasThreadPool() : nullptr;
 
-    const float* A = BufferA.GetBuffer(K * M);
+    float* A = BufferA.GetBuffer(K * M);
```
Comment (edgchen1) [inline onnxruntime/test/mlas/unittest/test_sqnbitgemm.cpp:201]: `const float* A`

File: onnxruntime/core/mlas/lib/sqnbitgemm.cpp [hunk 4]
```diff
@@ -11,38 +11,551 @@
...
+    const MLAS_SQNBIT_GEMM_DATA_PARAMS* DataParams,
+    void* PerGemmWorkspace,
+    size_t RangeStartM,
+    size_t RangeCountM,
+    size_t RangeStartN,
+    size_t RangeCountN
+);
+
+void
+SQNBitGemm_BlkBitWidth4_CompFp32(
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/sqnbitgemm.cpp:313]: >SQNBitGemm_BlkBitWidth4 [](http://example.com/codeflow?start=0&length=23)

nit: The prefix is too long. How about SQ4BitGemm_

File: onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp [hunk 5]
```diff
@@ -448,42 +414,344 @@
...
-#undef SPECIALIZE_QNBIT_BLK_DEQUANT_B_FOR_SGEMM
+    const std::byte* QuantA = QuantARowPtr;
+
+    const std::byte* QuantBData = QuantBDataColPtr;
+    const float* QuantBScale = QuantBScaleColPtr;
+    size_t QuantBZeroPointIdx = 0;  // track half byte increments with this index instead of a pointer
+
+    float32x4_t acc[NCols]{};
+
+    for (size_t k = 0; k < CountK; k += BlkLen) {
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp:566]: >for (size_t k = 0; k < CountK; k += BlkLen) { [](http://example.com/codeflow?start=4&length=45)

how is the leftover handled?

File: onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp [hunk 6]
```diff
@@ -448,42 +414,344 @@
...
+        int8_t b_zp[NCols];
+        if (QuantBZeroPointColPtr != nullptr) {
+            UnrolledLoop<NCols>([&](size_t i) {
+                const std::byte zp_packed =
+                    QuantBZeroPointColPtr[i * StrideQuantBZeroPoint + QuantBZeroPointIdx / 2];
+                b_zp[i] = ((QuantBZeroPointIdx & 1) == 1)
+                              ? std::to_integer<int8_t>(zp_packed >> 4)
+                              : std::to_integer<int8_t>(zp_packed & std::byte{0x0F});
+            });
+        } else {
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp:584]: this part have potential to be optimized depending on how the assembly is generated. It only needs to load from memory once to register for NCols <=8.

File: onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp (hunk 5 above)
Comment (edgchen1) [reply onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp:566]: A and B are both block quantized. the trailing elements in the A block beyond K should already be zeroed out during the A block quantization.

File: onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp [hunk 7]
```diff
@@ -448,42 +414,344 @@
...
+        }
+
+        for (size_t k_idx_in_blk = 0; k_idx_in_blk < k_blk_len; k_idx_in_blk += SubBlkLen) {
+            // load A row vector
+            int8x16_t av = vld1q_s8(a_data + k_idx_in_blk);
+
+            // load B column vectors
+            uint8x8_t bv_packed[NCols];
+            UnrolledLoop<NCols>([&](size_t i) {
+                const size_t b_data_block_offset = k_idx_in_blk * BlkBitWidth / 8;
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp:597]: > const size_t b_data_block_offset = k_idx_in_blk * BlkBitWidth / 8; [](http://example.com/codeflow?start=15&length=67)

Does the compiler optimize this out? It is independent with unroll loop?

Comment (edgchen1) [reply onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp:597]: good catch. moving it out does result in fewer assembly instructions in MSVC's output.

File: onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp (hunk 6 above)
Comment (edgchen1) [reply onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp:584]: the NCols zero point values are from different columns and not contiguous

**Review by yufenglee — APPROVED** [review]

File: onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp [hunk 8]
```diff
@@ -162,33 +164,27 @@ ComputeDotProducts(
...
             // load `SubBlkLen` elements from A, padded with 0's if there aren't enough
             const size_t k_subblk_len = std::min(k_blk_len - k_idx_in_blk, SubBlkLen);
             float32x4_t av[4]{};
-            LoadData<SubBlkLen>(ARowPtr + k + k_idx_in_blk, k_subblk_len, av);
+            LoadFloatData<SubBlkLen>(ARowPtr + k + k_idx_in_blk, k_subblk_len, av);
 
             // load B column vectors
             uint8x8_t bv_packed[NCols];
             UnrolledLoop<NCols>([&](size_t i) {
                 const size_t b_data_block_offset = k_idx_in_blk * BlkBitWidth / 8;
```
Comment (edgchen1) [inline onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp:178]: move this out of UnrolledLoop

**Review by yufenglee — APPROVED** [review]

**Pull request merged**